### ISSUE TITLE
Having `HOMEBREW_PREFIX/lib` in `LD_LIBRARY_PATH` breaks things when `glibc` installed

### ISSUE THREAD

**Issue by XuehaiPan** [Homebrew/homebrew-core#109893]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```console
linuxbrew@ubuntu:~$ brew config
HOMEBREW_VERSION: 3.5.10
ORIGIN: https://github.com/Homebrew/brew
HEAD: c5731faf8be1ddc1eeae4b162b88bd76318fc279
Last commit: 2 weeks ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 4c8e3e36cefc5f1afe56a8aae490f54d3e2b88bf
Core tap last commit: 2 weeks ago
Core tap branch: master
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 32
Homebrew Ruby: 2.6.8 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: 32-core 64-bit zen3
Clang: N/A
Git: 2.37.2 => /bin/git
Curl: 7.68.0 => /bin/curl
Kernel: Linux 5.10.102.1-microsoft-standard-WSL2 x86_64 GNU/Linux
OS: Unknown
Host glibc: 2.31
/usr/bin/gcc: 9.4.0
/usr/bin/ruby: N/A
glibc: N/A
gcc@5: N/A
xorg: N/A

linuxbrew@ubuntu:~$ brew doctor
Your system is ready to brew.
```

### Verification

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

I added the Homebrew lib path in my environment variables. Here I prepend `HOMEBREW_PREFIX/lib` into my `LD_LIBRARY_PATH`.

```bash
eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"
export C_INCLUDE_PATH="${HOMEBREW_PREFIX}/include${C_INCLUDE_PATH:+:"${C_INCLUDE_PATH}"}"
export CPLUS_INCLUDE_PATH="${HOMEBREW_PREFIX}/include${CPLUS_INCLUDE_PATH:+:"${CPLUS_INCLUDE_PATH}"}"
export LIBRARY_PATH="${HOMEBREW_PREFIX}/lib${LIBRARY_PATH:+:"${LIBRARY_PATH}"}"
export LD_LIBRARY_PATH="${HOMEBREW_PREFIX}/lib${LD_LIBRARY_PATH:+:"${LD_LIBRARY_PATH}"}"
```

### What happened (include all command output)?

Things break ("segmentation fault") after upgrading/installing `glibc` (if `HOMEBREW_PREFIX/lib` in `LD_LIBRARY_PATH`). This bug can be reproduced with both bottled and built-from-source `glibc`.

```console
linuxbrew@ubuntu:~$ export LD_LIBRARY_PATH="$(brew --prefix)/lib${LD_LIBRARY_PATH:+:"${LD_LIBRARY_PATH}"}"
linuxbrew@ubuntu:~$ echo "${LD_LIBRARY_PATH}"
/home/linuxbrew/.linuxbrew/lib

linuxbrew@ubuntu:~$ ls
Segmentation fault
linuxbrew@ubuntu:~$ mkdir
Segmentation fault
linuxbrew@ubuntu:~$ which
Segmentation fault
```

### What did you expect to happen?

Keep things work as expected when `HOMEBREW_PREFIX/lib` in `LD_LIBRARY_PATH`.

### Step-by-step reproduction instructions (by running `brew` commands)

```console
$ docker run -it --rm --pull=always -h ubuntu ghcr.io/homebrew/ubuntu20.04:latest
latest: Pulling from homebrew/ubuntu20.04
Digest: sha256:5fbcdc25290b1763e71ac8c4ea7cfbb9906686e3035d080768ae51f0fc8915c0
Status: Image is up to date for ghcr.io/homebrew/ubuntu20.04:latest
linuxbrew@ubuntu:~$ brew update

linuxbrew@ubuntu:~$ brew install glibc
==> Downloading https://ghcr.io/v2/homebrew/core/glibc/manifests/2.35
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/glibc/blobs/sha256:cd51b67e67b9a5d27880efc06a2b01b71e0b7b8b2cc207e6e961e46c423018e3
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:cd51b67e67b9a5d27880efc06a2b01b71e0b7b8b2cc207e6e961e46c423018e3?se=2022-09-07T13%3A15%3A00Z&sig=9CPHwMm7%2BMHon
######################################################################## 100.0%
==> Pouring glibc--2.35.x86_64_linux.bottle.tar.gz
==> Installing locale data for en_US.UTF-8
==> /home/linuxbrew/.linuxbrew/Cellar/glibc/2.35/bin/localedef -i en_US -f UTF-8 en_US.UTF-8
🍺  /home/linuxbrew/.linuxbrew/Cellar/glibc/2.35: 1,401 files, 47MB
==> Running `brew cleanup glibc`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).

linuxbrew@ubuntu:~$ ls
linuxbrew@ubuntu:~$ mkdir
mkdir: missing operand
Try 'mkdir --help' for more information.
linuxbrew@ubuntu:~$ which

linuxbrew@ubuntu:~$ export LD_LIBRARY_PATH="$(brew --prefix)/lib${LD_LIBRARY_PATH:+:"${LD_LIBRARY_PATH}"}"
linuxbrew@ubuntu:~$ echo "${LD_LIBRARY_PATH}"
/home/linuxbrew/.linuxbrew/lib

linuxbrew@ubuntu:~$ ls
Segmentation fault
linuxbrew@ubuntu:~$ mkdir
Segmentation fault
linuxbrew@ubuntu:~$ which
Segmentation fault
```

**Comment by carlocab** [comment]
CC @Homebrew/linux

**Comment by Bo98** [comment]
We should maybe consider not having `LD_LIBRARY_PATH` in `brew shellenv`.

It would be good to get any fix in for 3.6.0, but could have a 3.6.1 if necessary.

**Comment by carlocab** [comment]
> We should maybe consider not having `LD_LIBRARY_PATH` in `brew shellenv`.

We don't have `LD_LIBRARY_PATH` in `brew shellenv`.

**Comment by XuehaiPan** [comment]
Can we make `glibc` a keg-only formula and don't link it to `HOMEBREW_PREFIX/lib`?

**Comment by Bo98** [comment]
> > We should maybe consider not having `LD_LIBRARY_PATH` in `brew shellenv`.
> 
> We don't have `LD_LIBRARY_PATH` in `brew shellenv`.

In that case I'd recommend not setting it then in favour of proper linking and R(UN)PATHs

> Can we make `glibc` a keg-only formula and don't link it to `HOMEBREW_PREFIX/lib`?

Unless opt/glibc is in the RPATH of existing bottles, I don't think this can happen.

**Comment by XuehaiPan** [comment]
> Unless opt/glibc is in the RPATH of existing bottles, I don't think this can happen.

Only `gcc@x.y`, `llvm@x.y`, and `perl` uses `glibc`.

```console
$ grep -rlF '"glibc"' Formula | sort -V           
Formula/gcc.rb
Formula/gcc@4.9.rb
Formula/gcc@5.rb
Formula/gcc@6.rb
Formula/gcc@7.rb
Formula/gcc@8.rb
Formula/gcc@9.rb
Formula/gcc@10.rb
Formula/gcc@11.rb
Formula/llvm.rb
Formula/llvm@7.rb
Formula/llvm@8.rb
Formula/llvm@9.rb
Formula/llvm@11.rb
Formula/llvm@12.rb
Formula/llvm@13.rb
Formula/perl.rb
```

**Comment by carlocab** [comment]
> Unless opt/glibc is in the RPATH of existing bottles, I don't think this can happen.

It is not. It can probably be added at bottle-pour time.

> Only `gcc@x.y`, `llvm@x.y`, and `perl` uses `glibc`.

No. See `brew info hello` or `brew deps hello` with your local `brew` repository on the latest master in an Ubuntu 20.04 container.

**Comment by Bo98** [comment]
> > Unless opt/glibc is in the RPATH of existing bottles, I don't think this can happen.
> 
> Only `gcc@x.y`, `llvm@x.y`, and `perl` uses `glibc`.

As of 3.6.0, glibc will be a dependency of everything unless your host glibc is >= 2.35 (Ubuntu >= 22.04).

**Comment by sjackman** [comment]
```console
$ brew doctor
Your system is ready to brew.
```
`brew doctor` ought to complain if `LD_LIBRARY_PATH` contains `HOMEBREW_PREFIX/lib` and the brewed `glibc` is installed.

**Comment by danielnachun** [comment]
> ```
> $ brew doctor
> Your system is ready to brew.
> ```
> brew doctor ought to complain if LD_LIBRARY_PATH contains HOMEBREW_PREFIX/lib and the brewed glibc is installed.

Completely agreed here.  `glibc` will have a complete meltdown if you try to force the linker to use brewed `glibc` while still using the host `ld.so`.  You must either use the host `glibc` and `ld.so` together or the brewed `glibc` and `ld.so` together.

**Comment by carlocab** [comment]
Homebrew/brew#13826 -- untested, testers wanted

**Comment by ZhongRuoyu** [comment]
FYI: Having `HOMEBREW_PREFIX/include` in `CPATH` (or [other environment variables](http://gcc.gnu.org/onlinedocs/cpp/Environment-Variables.html) that tell `gcc` where to look for headers) also causes troubles when `glibc` is installed.

As an example:

```bash
[ruoyu@linuxbrew test]$ cat example.cc
#include <cmath>
int main() { double two = std::sqrt(4.0); }
[ruoyu@linuxbrew test]$ export CPATH=""
[ruoyu@linuxbrew test]$ g++ -o example example.cc  # compiles fine
[ruoyu@linuxbrew test]$ export CPATH="$(brew --prefix)/include"
[ruoyu@linuxbrew test]$ g++ -o example example.cc  # emits the following errors
In file included from /usr/include/math.h:289,
                 from /usr/include/c++/9/cmath:45,
                 from example.cc:1:
/home/ruoyu/opt/homebrew/include/bits/mathcalls-helper-functions.h:20:24: error: '__fpclassify' has not been declared
   20 | __MATHDECL_ALIAS (int, __fpclassify,, (_Mdouble_ __value), fpclassify)
      |                        ^~~~~~~~~~~~
/home/ruoyu/opt/homebrew/include/bits/mathcalls-helper-functions.h:20:37: error: expected identifier before ',' token
   20 | __MATHDECL_ALIAS (int, __fpclassify,, (_Mdouble_ __value), fpclassify)
      |                                     ^
/home/ruoyu/opt/homebrew/include/bits/mathcalls-helper-functions.h:20:39: error: expected identifier before '(' token
   20 | __MATHDECL_ALIAS (int, __fpclassify,, (_Mdouble_ __value), fpclassify)
      |                                       ^
/home/ruoyu/opt/homebrew/include/bits/mathcalls-helper-functions.h:20:60: error: 'fpclassify' has not been declared
   20 | __MATHDECL_ALIAS (int, __fpclassify,, (_Mdouble_ __value), fpclassify)
      |                                                            ^~~~~~~~~~
/home/ruoyu/opt/homebrew/include/bits/mathcalls-helper-functions.h:21:33: error: expected constructor, destructor, or type conversion before ';' token
   21 |      __attribute__ ((__const__));
      |                                 ^
# ... and more errors follow
```

I am running this on Ubuntu 20.04, which ships `glibc` version 2.31.

**Comment by sjackman** [comment]
> I am running this on Ubuntu 20.04, which ships glibc version 2.31. I wonder if this means I'm not supposed to do this?

@ZhongRuoyu Is upgrading to Ubuntu 22.04, which provides Glibc 2.35, an option for you? You wouldn't need to install the brewed `glibc` if so.

**Comment by ZhongRuoyu** [comment]
> > I am running this on Ubuntu 20.04, which ships glibc version 2.31. I wonder if this means I'm not supposed to do this?
> 
> @ZhongRuoyu Is upgrading to Ubuntu 22.04, which provides Glibc 2.35, an option for you? You wouldn't need to install the brewed `glibc` if so.

@sjackman Unfortunately, no, as I don't have root access to that machine.

**Comment by sjackman** [comment]
PR https://github.com/Homebrew/homebrew-core/pull/109998 may resolve this issue.
- https://github.com/Homebrew/homebrew-core/pull/109998

**Comment by XuehaiPan** [comment]
Resolved by Homebrew/brew#13873.
